tools/clang/lib/Analysis/UninitializedValues.cpp

```diff
@@ -824,6 +885,37 @@ static bool runOnBlock(const CFGBlock *block, const CFG &cfg,
     if (Optional<CFGStmt> cs = I->getAs<CFGStmt>())
       tf.Visit(const_cast<Stmt*>(cs->getStmt()));
   }
+
+  // HLSL Change Begin - Treat `out` parameters as uninitialized values.
+  // If this block has no successors and does not have a terminator stmt which
+  // we would have handled above.
+  if (block->succ_size() == 0 && !block->getTerminator()) {
```

Comment:
This case occurs when a function (usually `void`) doesn't have an explicit return statement at the end of the function. It isn't strictly related to `noreturn`. This is an extremely common case where out parameters are used, so omitting it would significantly limit the usefulness of the diagnostic.

Reply:
I was trying simplify reporting for this case (but didn't add it to the tests):

```c++
void fn(out float F, bool B) {
  if(B)
    return;
  else
    return;
```

The idea being if all control flow blocks that fall off the end of the function have explicit return statements we shouldn't separately diagnose the end of the function.

Reply:
Adding a test for this, it is actually important. Take this example:

```c++
void Fn(out Float F, bool B) {
  if (B) {
    F = 1.0;
    return;
  } else {
    F = 2.0;
    return;
}
```

We should not emit any diagnostic here.